File: services/content/internal/objectstore/objectstore.go
// Package objectstore wraps the MinIO/S3 client content uses to presign
// read-only URLs for summary-audio objects it owns the metadata of but
// never stores the bytes of, per ADR 0001. Read-only: unlike media's
// objectstore, content never writes or deletes objects — the worker owns
// the write, content only signs a GET for playback.
package objectstore

import (
	"context"
	"fmt"
	"net/url"
	"time"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

// Client presigns GET URLs for objects in one bucket.
type Client struct {
	minio  *minio.Client
	bucket string
}

// New returns a Client for the bucket on the given MinIO/S3-compatible
// endpoint.
func New(endpoint, accessKey, secretKey string, useSSL bool, bucket string) (*Client, error) {
	c, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
		Secure: useSSL,
	})
	if err != nil {
		return nil, fmt.Errorf("new minio client: %w", err)
	}
	return &Client{minio: c, bucket: bucket}, nil
}

// PresignGetObject returns a short-lived URL for reading objectKey
// directly from object storage.
func (c *Client) PresignGetObject(ctx context.Context, objectKey string, expires time.Duration) (string, error) {
	u, err := c.minio.PresignedGetObject(ctx, c.bucket, objectKey, expires, url.Values{})
	if err != nil {
		return "", fmt.Errorf("presign get object: %w", err)
	}
	return u.String(), nil
}

File: services/worker/src/clients/objectstore.py
"""MinIO client for the write side of the derivative/summary-audio flow:
worker uploads the object, then registers its key through media or content
(never the reverse — object storage never stores generated text, per
docs/architecture.md's storage boundary).
"""

from __future__ import annotations

import io

from minio import Minio


class ObjectStore:
    def __init__(self, *, endpoint: str, access_key: str, secret_key: str, secure: bool, bucket: str) -> None:
        self._client = Minio(endpoint=endpoint, access_key=access_key, secret_key=secret_key, secure=secure)
        self._bucket = bucket

    def put_object(self, object_key: str, data: bytes, content_type: str) -> str:
        """Writes data under object_key and returns it unchanged, so
        callers can chain straight into media.RegisterDerivative or
        content.StoreSummaryAudioMetadata.
        """
        self._client.put_object(
            bucket_name=self._bucket,
            object_name=object_key,
            data=io.BytesIO(data),
            length=len(data),
            content_type=content_type,
        )
        return object_key
